xonsh/completer.py

```diff
@@ -78,16 +78,54 @@ def _normpath(p):
 
     return p
 
-class Completer(object):
-    """This provides a list of optional completions for the xonsh shell."""
+class KnownCommands(object):
 
     def __init__(self):
-        # initialize command cache
         self._path_checksum = None
         self._alias_checksum = None
         self._path_mtime = -1
         self._cmds_cache = frozenset()
+    
+    def all_commands(self):
+        """find all commands available in PATH
+        """
+        path = builtins.__xonsh_env__.get('PATH', [])
+        # did PATH change?
+        path_hash = hash(tuple(path))
+        cache_valid = path_hash == self._path_checksum
+        self._path_checksum = path_hash
+        # did aliases change?
+        al_hash = hash(tuple(sorted(builtins.aliases.keys())))
```

Comment:
While we're making changes here, I think we can get a speedup here if, instead of sorting, tuple-izing, and hashing, we hash a frozenset of the keys instead:

That is, if:
```
hash(tuple(sorted(builtins.aliases.keys())))
```
becomes
```
hash(frozenset(builtins.aliases.keys()))
```

Not sure just how significant the speedup will be (quick hacky tests, I get about a factor of 2 on my machine).